mllib/src/main/scala/org/apache/spark/ml/tree/impl/RandomForest.scala

```diff
@@ -1009,10 +1009,24 @@ private[spark] object RandomForest extends Logging {
       // sort distinct values
       val valueCounts = valueCountMap.toSeq.sortBy(_._1).toArray
 
-      // if possible splits is not enough or just enough, just return all possible splits
+      // perhaps weighted mean is better in the future, see SPARK-16957 and Github PR 17556.
+      def mean(pre: (Double, Int), cur: (Double, Int)): Double = {
+        val (preValue, preCount) = pre
```

Comment:
Is it worth factoring a method for this? you could just write `(preValue, _) =` here, but, just dereferncing `._1` isn't so bad, and then, wondering if it saves much to make a method.